src/dune/lib.ml

```diff
@@ -1490,89 +1523,119 @@ end = struct
-              | Status.Found t' ->
-                if t = t' then
+              match find_internal db lib.name ~stack with
+              | Status.Found lib' ->
+                if lib = lib' then
                   Ok ()
                 else
                   let req_by =
-                    Dep_stack.to_required_by stack ~stop_at:orig_stack
+                    Dep_stack.to_required_by stack ~stop_at:t.orig_stack
                   in
-                  Error.overlap ~in_workspace:t'.info ~installed:(t.info, req_by)
+                  Error.overlap ~in_workspace:lib'.info
+                    ~installed:(lib.info, req_by)
               | _ -> assert false )
           in
-          let* new_stack = Dep_stack.push stack ~implements_via (to_id t) in
-          let* deps = t.requires in
-          let* unimplemented' = Vlib.Unimplemented.add !unimplemented t in
-          unimplemented := unimplemented';
+          let* new_stack = Dep_stack.push stack ~implements_via (to_id lib) in
+          let* deps = lib.requires in
+          let* unimplemented' = Vlib.Unimplemented.add t.unimplemented lib in
+          t.unimplemented <- unimplemented';
           let+ () =
-            Result.List.iter deps ~f:(fun l -> loop (None, l) ~stack:new_stack)
+            Result.List.iter deps ~f:(fun l ->
+                visit t (None, l) ~stack:new_stack)
           in
-          res := (t, stack) :: !res
+          t.result <- (lib, stack) :: t.result
+  end
+
+  let step1_closure db ts ~stack:orig_stack ~forbidden_libraries =
+    let state = Closure.make ~db ~forbidden_libraries ~orig_stack in
+    let+ () =
+      Result.List.iter ts ~f:(fun lib ->
+          Closure.visit state ~stack:orig_stack (None, lib))
     in
-    (* Closure loop with virtual libraries/variants selection*)
-    let rec handle ts ~stack =
-      let* () = Result.List.iter ts ~f:(loop ~stack) in
-      if not linking then
-        Ok ()
-      else
-        let* impls =
-          let+ impls =
-            Vlib.Unimplemented.implemented_via_variants !unimplemented ~variants
-          in
-          List.map impls ~f:(fun (via, impl) -> (Some via, impl))
-        in
-        match impls with
-        | _ :: _ -> handle impls ~stack
-        | [] -> (
-          let* defaults =
-            let default_impls =
-              Vlib.Unimplemented.with_default_implementations !unimplemented
-            in
-            resolve_default_libraries default_impls ~variants
-          in
-          match defaults with
-          | [] -> Ok ()
-          | _ :: _ ->
-            let defaults =
-              List.map defaults ~f:(fun (via, lib) -> (Some via, lib))
-            in
-            handle defaults ~stack )
+    state
+
+  let compile_closure_with_overlap_checks db ts ~stack ~forbidden_libraries =
+    let* state = step1_closure db ts ~stack ~forbidden_libraries in
+    Closure.result state ~linking:false
+
+  let linking_closure_with_overlap_checks db ts ~stack ~variants
```

Comment:
Because we've introduced this `Closure` module, we're now able to separate the compilation from the linking step. Previously, the two behaviors were controlled with a `~linking` flag.